src/game/client/worldheightmap.cpp

```diff
@@ -29,21 +45,2358 @@ class GDIFileStream : public InputStream
+                break;
+            }
+        }
+
+        if (b) {
+            int x = 72 * l + 4;
+            int y = 72 * j + 4;
+            this->m_edgeTextureClasses[i].position_in_texture.x = x;
+            this->m_edgeTextureClasses[i].position_in_texture.y = y;
+
+            if (i3 < y + (width << 6) + 4) {
+                i3 = y + (width << 6) + 4;
+            }
+
+            for (int m = 0; m < width; m++) {
+                for (int n = 0; n < width; n++) {
+                    availableGrid[j][28 * n + m + l] = 0;
+                    int i2 = width * n + m + this->m_edgeTextureClasses[i].first_tile;
+
+                    if (this->m_edgeTiles[i2]) {
+                        this->m_edgeTiles[i2]->m_tileLocationInTexture.x = (m << 6) + x;
+                        this->m_edgeTiles[i2]->m_tileLocationInTexture.y = ((width - n - 1) << 6) + y;
+                    }
+                }
+            }
+        } else {
+            this->m_edgeTextureClasses[i].position_in_texture.x = 0;
+            this->m_edgeTextureClasses[i].position_in_texture.y = 0;
+        }
+    }
+
+    if (edge_height) {
+        *edge_height = i3;
+    }
+
+    return i1;
+}
+
+void WorldHeightMap::Get_UV_For_Ndx(int tile_ndx, float *min_u, float *max_u, float *min_v, float *max_v, bool full_tile)
+{
+    if (m_sourceTiles[tile_ndx >> 2]) {
+        TileData *tile = m_sourceTiles[tile_ndx >> 2];
+        *min_u = (float)tile->m_tileLocationInTexture.x;
+        *max_u = (float)tile->m_tileLocationInTexture.y;
+        *min_v = *min_u + 64.0f;
+        *max_v = *max_u + 64.0f;
+        *min_u = *min_u / 2048.0f;
+        *max_u = *max_u / (float)m_terrainTexHeight;
+        *min_v = *min_v / 2048.0f;
+        *max_v = *max_v / (float)m_terrainTexHeight;
+
+        if (!full_tile) {
+            float min = (*min_u + *min_v) / 2.0f;
+            float max = (*max_u + *max_v) / 2.0f;
+
+            if ((tile_ndx & 2) != 0) {
+                *max_v = max;
+            } else {
+                *max_u = max;
+            }
+
+            if ((tile_ndx & 1) != 0) {
+                *min_u = min;
+            } else {
+                *min_v = min;
+            }
+        }
+    } else {
+        *max_v = 0.0f;
+        *min_v = 0.0f;
+        *max_u = 0.0f;
+        *min_u = 0.0f;
+    }
+}
+
+bool WorldHeightMap::Is_Cliff_Mapped_Texture(int x, int y)
+{
+    int i = (m_drawOriginY + y) * m_width + m_drawOriginX + x;
+    return i >= 0 && i < m_dataSize && m_cliffInfoNdxes[i] != 0;
+}
+
+bool WorldHeightMap::Get_UV_Data(int x_index, int y_index, float *const u, float *const v, bool full_tile)
+{
+    int i = m_drawOriginX + x_index + m_width * (m_drawOriginY + y_index);
+
+    if (i < m_dataSize && m_tileNdxes) {
+        return Get_UV_For_Tile_Index(i, m_tileNdxes[i], u, v, full_tile);
+    } else {
+        return false;
+    }
+}
+
+bool WorldHeightMap::Get_UV_For_Tile_Index(int ndx, short tile_ndx, float *const v, float *const u, bool full_tile)

```

Comment:
because Get_UV_For_Ndx is wrong this will likely be too

Reply:
fixed.